src/components/tx-flow/index.tsx

```diff
@@ -19,68 +17,64 @@ export const TxModalContext = createContext<TxModalContextType>({
+const confirmClose = () => {
+  return confirm('Closing this window will discard your current progress.')
+}
 
 export const TxModalProvider = ({ children }: { children: ReactNode }): ReactElement => {
   const [txFlow, setFlow] = useState<TxModalContextType['txFlow']>(undefined)
-  const [shouldWarn, setShouldWarn] = useState<boolean>(false)
-  const [, setOnClose] = useState<Parameters<TxModalContextType['setTxFlow']>[1]>(noop)
   const [fullWidth, setFullWidth] = useState<boolean>(false)
-  const pathname = usePathname()
-  const [, setLastPath] = useState<string>(pathname)
-  const { safeAddress, safe } = useSafeInfo()
+  const shouldWarn = useRef<boolean>(true)
+  const onClose = useRef<() => void>(noop)
+  const searchParams = useSearchParams()
+  const safeQuery = searchParams.get('safe')
+  const prevQuery = useRef<string>(safeQuery ?? '')
 
   const handleModalClose = useCallback(() => {
-    setOnClose((prevOnClose) => {
-      prevOnClose?.()
-      return noop
-    })
+    onClose.current()
+    onClose.current = noop
     setFlow(undefined)
-    txDispatch(TxEvent.USER_QUIT, {})
-  }, [setFlow, setOnClose])
+  }, [])
 
   const handleShowWarning = useCallback(() => {
-    if (!shouldWarn) {
-      handleModalClose()
+    if (shouldWarn.current && !confirmClose()) {
       return
     }
-
-    if (!shouldClose()) return
-
     handleModalClose()
-  }, [shouldWarn, handleModalClose])
+  }, [handleModalClose])
 
   const setTxFlow = useCallback(
-    (newTxFlow: TxModalContextType['txFlow'], onClose?: () => void, newShouldWarn?: boolean) => {
-      // If flow is open and user opens a different one, show confirmation dialog if required
-      if (txFlow && newTxFlow && newTxFlow?.type !== SuccessScreen) {
-        if (shouldWarn && !shouldClose()) return
-        txDispatch(TxEvent.USER_QUIT, {})
-      }
+    (newTxFlow: TxModalContextType['txFlow'], newOnClose?: () => void, newShouldWarn?: boolean) => {
+      setFlow((prev) => {
+        if (prev === newTxFlow) return prev
+
+        // If a new flow is triggered, close the current one
+        if (prev && newTxFlow?.type !== SuccessScreen) {
+          if (shouldWarn.current && !confirmClose()) {
+            return prev
+          }
+          onClose.current()
+        }
 
-      setFlow(newTxFlow)
-      setOnClose(() => onClose ?? noop)
-      setShouldWarn(newShouldWarn ?? true)
+        onClose.current = newOnClose ?? noop
+        shouldWarn.current = newShouldWarn ?? true
+
+        return newTxFlow
+      })
     },
-    [txFlow, shouldWarn],
+    [],
   )
 
-  // Show the confirmation dialog if user navigates
+  // // Close the modal when the Safe changes
   useEffect(() => {
-    setLastPath((prev) => {
-      if (prev !== pathname && txFlow) {
-        handleShowWarning()
-      }
-      return pathname
-    })
-  }, [txFlow, handleShowWarning, pathname])
+    if (safeQuery === prevQuery.current) return
+    prevQuery.current = safeQuery ?? ''
```

Comment:
This is not taking into account anymore if the user navigates to a different route within the same safe. The tx-flow now stays open instead of showing the user a warning.

Reply:
As long as they are in the same Safe, navigating shouldn't matter, should it?

Reply:
Good point, I'll restore it.